### ISSUE TITLE
Playlist sorting by Filename / filename (without path) erroneously

### ISSUE THREAD

**Issue by kekkc** [clementine-player/Clementine#5445]
### Before posting

Please follow the steps below and check the boxes with [x] once you did the step.

- [X] I checked the issue tracker for similar issues
- [X] I checked the [changelog](https://github.com/clementine-player/Clementine/blob/master/Changelog) if the issue is already resolved
- [X] I tried the latest Clementine build from [here](https://builds.clementine-player.org/)

### System information

Please provide information about your system and the version of Clementine used.

- Operating System: Win (OS independent)
- Clementine version: 1.3.1

### Expected behaviour / actual behaviour
just discovered Clementine (amazing player and much smaller on Windows than Amarok), but I realized a strange behavior when new files are added to the playlist. If a folder "My Music" is added that contains a folder "Brooklyn", Clementine displays it as the following:

1. "My Music" songs starting with "A" are displayed first
2. "My Music" songs starting with "B" are displayed afterwards
3. "Brooklyn" songs starting with "A" are displayed afterwards
4. "My Music" songs starting with "C" are displayed afterwards

Usually players like Winamp, Foobar etc.. will display logically sorted based on the folder structure:

1. "My Music" songs starting with "A" are displayed first
2. "My Music" songs starting with "B" are displayed afterwards
3. "My Music" songs starting with "C" are displayed afterwards (and all remaining files in "My Music")
4. "Brooklyn" songs starting with "A" are displayed afterwards

All files within a folder are displayed first, then the subfolders files & folders are displayed

### Steps to reproduce the problem (only for bugs)
1. Add & folder with subfolders to the playlist
2. Sort by filename

**Comment by santigl** [comment]
Thanks for reporting this bug. I can confirm that it is not working properly.

So far I could think of a solution that would return a *breadth-first* order, at least implemented the way sorting functions are now (by defining a less-than comparison function).

A sample output would be:
* /Music/a.mp3
* /Music/c.mp3
* /Music/Brooklyn/b.mp3
* /Music/Brooklyn/z.mp3
* /Music/Charlie/a.mp3
* /Music/Charlie/e.mp3
* /Music/Brooklyn/X/l.mp3

This takes into account the "depth level" of a file, and as you can see doesn't scan recursively a given directory (that would be a *depth-first* ordering). So files in subdirectories under `/music/Brooklyn/`, in this case `/music/Brooklyn/X/.mp3` (level 3) appear after all files under level 2 (in this case, `/Music/Charlie/*.mp3`).

Would this order make any sense? 

The implementation is super simple. It consists on doing a second pass stable-sorting the files by depth levels.

Does anybody know of an algorithm that would determine the relative hierarchy of two paths given them as strings? Any other alternatives would imply changing the structure of the column-sorting logic.

**Comment by kekkc** [comment]
Awesome, many thanks for the quick patch ;)

>EDIT: Just for some extra information, I checked VLC and it sorts by lexicographical order. So we get this:
This VLC behavior seems logic for me. It parses the first path level, then it parses the 2nd one completely (incl. subfolders) and so on.

**Comment by santigl** [comment]
Yeah, and it's easier to understand its logic at a glimpse. The alternative is less intuitive.

**Comment by kekkc** [comment]
Any chance that this will be implemented soon, or are there any issues hindering landing of this patch?

**Comment by santigl** [comment]
It was merged to the master branch, so it should be on future releases. Meanwhile, you can compile it from source or download a binary from the Buildbot.

**Comment by kekkc** [comment]
Thanks, but I think sorting didn't change. I tested it with v1.3.1-268-g789c492 from 27.12..

- /Music/a.mp3
- /Music/Charlie/a.mp3
- /Music/Brooklyn/b.mp3
- /Music/c.mp3
- /Music/Charlie/e.mp3
- /Music/Brooklyn/X/l.mp3
- /Music/Brooklyn/z.mp3

Strangely there are some exceptions when sorting by filename. I get some songs with "z" in the middle, then some others with "t" from the top root folder and lastly some others starting with "z". Seems not to be related to case sensitivity, but I was not able to see the "logic" behind it. Please let me know if you need some more screens, would be interesting if it works for you or if it's just some weird difference between Win & Lux.
![unbenannt](https://cloud.githubusercontent.com/assets/6283064/21556914/54445ed2-ce25-11e6-93d7-4a4a32acc207.png)

**Comment by santigl** [comment]
You're right! I just realized that the patch that was merged was the one to make the comparison case-insensitive. But the one that would change the logic wasn't (and I can't recall if there was any reason why).

I am going to ask, and I'll get back to you.

**Comment by santigl** [comment]
Well, apparently the patch was good. I might have been waiting for confirmation before submitting it, and then I never got around to do it. Thanks a lot for the reminder!

It is now merged into master (thanks to @hatstand). You can get the latest Windows build [here](http://builds.clementine-player.org/win32/release/ClementineSetup-1.3.1-269-g00c96f7.exe). When you do, would you let me know if it worked for you?

**Comment by kekkc** [comment]
Happy new year ;)

Many thanks for the update. I tested it and all files in root are sorted on the top, very cool.

I just noticed a small glitch (hard to explain, made a test file again):
[Music.zip](https://github.com/clementine-player/Clementine/files/680000/Music.zip)

Please notice the folder "alben". This contains one song and several subfolders. So currently Clementine sorts the one song at the top, but afterwards it doesn't display the subfolders of the current "alben" folder, but is continuing with the subfolders from the root folder (one level up). After those are displayed, Clementine lastly includes the "alben" subfolders as the last part.

BTW: propably known, but searching the web I also found some link for sorting in QT (seems logical sorting was just included with QT5.2): http://doc.qt.io/qt-5/qcollator.html .
